lib/puppet/parser/catalog_validator.rb

```diff
@@ -0,0 +1,33 @@
+# Abstract class for a catalog validator that can be registered with the compiler to run at
+# a certain stage.
+module Puppet::Parser
```

Comment:
The Compiler is in Parser namespace (which admittedly is a bit odd). Validation is performed by the compiler. In my view, the catalog is the result of compilation and the validation is a part of that.